### PR TITLE
Implement application command permissions v2

### PR DESCRIPTION
**Pull request by BobDotCom**
Implements the new permissions system.
Original PR: #1129

### CODE REVIEW

**Comment by TurnrDev** [comment]
Just to be clear, what will this PR bring? What are the new changes to the permissions system?

**Comment by Dorukyum** [comment]
This allows bot devs server managers to choose permission requirements (such as "manage messages") for commands

**Comment by AmirAvram** [comment]
Will permissions v2 deprecate the ability to allow specific user_ids for commands? or is it just not planned for next release?

**Comment by TurnrDev** [comment]
Is this PR frozen? Is there anything I can do to contribute?

**Comment by Lulalaby** [comment]
> Is this PR frozen? Is there anything I can do to contribute?

Bob is atm out of order and the permission thing is not yet public

**Comment by krittick** [comment]
Perms v2 is now available to all servers as of today: https://github.com/discord/discord-api-docs/pull/4830/files

File: discord/bot.py [hunk 1]
```diff
@@ -559,6 +560,8 @@ async def sync_commands(
                     guild_commands, guild_id=guild_id, force=force
                 )
 
+        return
+
         # TODO: 2.1: Remove this and favor permissions v2
         # Global Command Permissions
         global_permissions: List = []
```
Comment (plun1331) [inline discord/bot.py:567]: I assume this code can be removed now

File: docs/api.rst [hunk 2]
```diff
@@ -186,29 +186,6 @@ AutocompleteContext
 .. autoclass:: AutocompleteContext
     :members:
 
-CommandPermission
```
Comment (plun1331) [inline docs/api.rst:189]: New permission decorators should be documented

**Review by plun1331 — COMMENTED** [review]
I'll work on a PR to add the new stuff documented in the Discord API PR later today, if it's not already being worked on

File: discord/bot.py (hunk 1 above)
Comment (krittick) [reply discord/bot.py:649]: Yup.

**Comment by krittick** [comment]
> I'll work on a PR to add the new stuff documented in the Discord API PR later today, if it's not already being worked on

Sounds great, thank you! I just resolved merge conflicts as well.

**Comment by Lulalaby** [comment]
That's Teamwork ❤️

File: discord/commands/core.py [hunk 3]
```diff
@@ -1619,42 +1624,42 @@ def validate_chat_input_name(name: Any, locale: Optional[str] = None):
     # Must meet the regex ^[\w-]{1,32}$
     if locale is not None and locale not in valid_locales:
         raise ValidationError(
-            f"Locale '{locale}' is not a valid locale, "
-            f"see {docs}/reference#locales for list of supported locales."
+            f"Locale '{locale}' is not a valid locale, " f"see {docs}/reference#locales for list of supported locales."
```
Comment (BobDotCom) [inline discord/commands/core.py:1627]: What happened here

File: discord/commands/core.py [hunk 4]
```diff
@@ -1619,42 +1624,42 @@ def validate_chat_input_name(name: Any, locale: Optional[str] = None):
...
+            error.args = (f"{error.args[0]} in locale {locale}",)
         raise error
 
 
 def validate_chat_input_description(description: Any, locale: Optional[str] = None):
     if locale is not None and locale not in valid_locales:
         raise ValidationError(
-            f"Locale '{locale}' is not a valid locale, "
-            f"see {docs}/reference#locales for list of supported locales."
+            f"Locale '{locale}' is not a valid locale, " f"see {docs}/reference#locales for list of supported locales."
```
Comment (BobDotCom) [inline discord/commands/core.py:1652]: What happened here

**Review by Lulalaby — DISMISSED** [review]

File: discord/commands/core.py [hunk 5]
```diff
@@ -1624,9 +1624,9 @@ def validate_chat_input_name(name: Any, locale: Optional[str] = None):
             f"Locale '{locale}' is not a valid locale, " f"see {docs}/reference#locales for list of supported locales."
         )
     error = None
-    if not isinstance(name, str):
-        error = TypeError(f'Command names and options must be of type str. Received "{name}"')
-    elif not re.match(r"^[\w-]{1,32}$", name):
+    if not isinstance(name, str): not re.match(r"^[\w-]{1,32}$", name):
```
Comment (TurnrDev) [inline discord/commands/core.py:1627]:
```suggestion
    if not isinstance(name, str) or not re.match(r"^[\w-]{1,32}$", name):
```

**Review by TurnrDev — CHANGES_REQUESTED** [review]

**Review by Lulalaby — APPROVED** [review]

File: discord/commands/core.py [hunk 6]
```diff
@@ -88,6 +87,9 @@
     from typing_extensions import Concatenate, ParamSpec
 
     from ..cog import Cog
+    from .. import Permissions
+else:
+    Permissions = TypeVar("Permissions")
```
[the remark this answers is not in the document]
Comment (TurnrDev) [reply discord/commands/core.py:92]: #1324

**Review by Lulalaby — DISMISSED** [review]

**Comment by plun1331** [comment]
Pretty much done except for gateway events

**Review by Lulalaby — APPROVED** [review]

**Review by krittick — APPROVED** [review]

**Comment by Lulalaby** [comment]
@krittick wanna resolve mr conflicts?

**Comment by krittick** [comment]
> @krittick wanna resolve mr conflicts?

Working on it now.

**Pull request merged**